table/tables/tables.go

```diff
@@ -416,12 +416,15 @@ func (t *Table) AddRecord(ctx context.Context, r []interface{}) (recordID int64,
 	if t.meta.PKIsHandle {
 		// Check key exists.
 		recordKey := t.RecordKey(recordID, nil)
+		e := kv.ErrKeyExists.Gen("Duplicate entry '%d' for key 'PRIMARY'", recordID)
+		txn.SetOption(kv.PresumeKeyNotExistsError, e)
 		_, err = txn.Get(recordKey)
 		if err == nil {
-			return recordID, kv.ErrKeyExists
+			return recordID, errors.Trace(e)
 		} else if !terror.ErrorEqual(err, kv.ErrNotExist) {
 			return 0, errors.Trace(err)
 		}
+		txn.DelOption(kv.PresumeKeyNotExistsError)
```

Comment:
I think it is better to use a `defer` to delete option, or the options value will be left undeleted if error occurs.